### PR TITLE
Mutation addition and minor updates

### PR DESCRIPTION
**Pull request by Vanquish59**
Added mutation support.
Small logic updates.

### CODE REVIEW

File: src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java [hunk 1]
```diff
@@ -358,6 +386,28 @@ private boolean isLeaf(Node node) {
         return node.getChildren().isEmpty();
     }
 
+    public boolean isMutation(String query){
+        return query.contains(MUTATION_DEFINITION);
```
Comment (vaant) [inline src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java]: This appears insufficient, what if there is string with the value "mutation" somewhere in the query?

Comment (Vanquish59) [reply src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java]: So we have two other choices.

1) check if the initial string contains 'mutation'
2) Create a document and verify through there.

File: src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java [hunk 2]
```diff
@@ -60,39 +65,46 @@ public int checkAndRetrieveQueryScore(String query) {
...
-     * @return the length of the query string.
+     * @param query - graphql query string
+     * @return the length of the query string.  If the query is a mutation, we retrieve the return query body of the mutation
      */
     public int characterScore(String query) {
         if (StringUtils.isBlank(query)) {
             throw new GlitrException("query cannot be null or empty");
         }
 
+        if(isMutation(query)){
```
Comment (vaant) [inline src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java]: add a space after the if

File: src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java [hunk 3]
```diff
@@ -268,30 +292,34 @@ private int queryScore(Node queryNode, int depth) {
...
+     * @param query - graphql query string
      * @return the root node.  We parse the query with a graphql library and retrieve a document object.  We filter the
-     * document object in search of an OPERATION_DEFINITION and make sure its a QUERY and not a MUTATION.  Once we find
-     * the object, we return it.
+     * document object in search of an OPERATION_DEFINITION.  If the query is a MUTATION we extract the return query
+     * from the mutation body.  If it's not a mutation, we parse as is.
      */
-    private Node parseRootNode(String query) {
+    private OperationDefinition parseRootNode(String query) {
+
```
Comment (vaant) [inline src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java]: remove newline

File: src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java [hunk 4]
```diff
@@ -60,39 +65,46 @@ public int checkAndRetrieveQueryScore(String query) {
             throw new GlitrException(String.format("query depth has exceeded the maximum depth level of %d.", maxDepthLimit));
         }
 
-        return queryScore(query);
+        if(scoreLimitExceeded(query)){
```
Comment (vaant) [inline src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java]: add space after the if

File: src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java (hunk 1 above)
Comment (vaant) [reply src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java]: yeah, you should just check that it starts with "mutation" (ignoring whitespace). make sure to adds tests around this though.

Comment (Vanquish59) [reply src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java]: cool thanks

File: src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java [hunk 5]
```diff
@@ -358,6 +385,28 @@ private boolean isLeaf(Node node) {
...
+    public boolean isMutation(String query){
+        return query.trim().startsWith(MUTATION_DEFINITION) || query.trim().startsWith(StringUtils.capitalize(MUTATION_DEFINITION));
+    }
+
+    /**
+     * @param query - graphql query string
+     * @return the query that is returned by the mutation response body
+     */
+    private String extractReturnQueryFromMutation(String query){
+
```
Comment (vaant) [inline src/main/java/com/nfl/glitr/util/QueryComplexityCalculator.java]: remove newline

**Comment by tinnou** [comment]
Cool to see you are working on query complexity! I believe the main issue with this approach is it won't reflect the actual data fetching complexity of the underlying system. One field might be very expensive to fetch, and there is no way for the developer to pass a complexity function/value that can affect the total score in a significant way.

Did you guys look into how Sangria is doing it? https://github.com/sangria-graphql/sangria/blob/9a5aed829d977badc769f17310357ac464039a63/src/main/scala/sangria/execution/QueryReducer.scala#L65. Here is a pointer to the documentation http://sangria-graphql.org/learn/#verification-with-query-reducers.

**Comment by vaant** [comment]
@tinnou hey man, yeah, finally getting back around to it. Decided to start with a simple implementation and iterate rather than working off a full feature set. I do like the notion of a query reducer, will need to look into it a bit more.

What I was thinking of in the next evolution was adding a new annotation to influence query complexity along with query complexity becoming deeply integrated into glitr initialization (and thus able to leverage our reflection suite). This would essentially be an expansion of TypeRegistry to hold a per field complexity multipler. Thoughts?

**Comment by tinnou** [comment]
Yea the annotation should work, maybe it would take a value at the beginning and we can later have it take an expression. Something similar to SpEL, i.e `@QueryComplexity("25.0D + #{args.limit} * #{childScore}")`

**Pull request merged**